File: examples/x3_oop_aliens.py
#!/usr/bin/env python3
'''
@author Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d
from x2_oop_alien import Alien

def tick():
    g2d.clear_canvas()
    for a in actors:
        a.move()
        g2d.fill_rect(a.position())

actors = [Alien((40, 40)), Alien((80, 80)), Alien((120, 40))]
g2d.init_canvas((480, 360))
g2d.main_loop(tick)

File: go/x3_oop_aliens.go
package main

import . "g2d"

var actors = []*Alien { NewAlien(Point{40, 40}),
    NewAlien(Point{80, 80}), NewAlien(Point{120, 40}) }

type Alien struct {
    x, y, w, h int
    xmin, xmax int
    dx, dy     int
}

func NewAlien(pos Point) *Alien {
    return &Alien{pos.X, pos.Y, 20, 20, pos.X, pos.X+150, 5, 5}
}

func (a *Alien) Move() {
    if a.xmin <= a.x+a.dx && a.x+a.dx <= a.xmax {
        a.x += a.dx
    } else {
        a.dx = -a.dx
        a.y += a.dy
    }
}

func (a *Alien) Position() Point {
    return Point{a.x, a.y}
}

func tick() {
    ClearCanvas()
    for _, a := range actors {
        a.Move()
        DrawImage("ball.png", a.Position())
    }
}

func main() {
    InitCanvas(Point{480, 360})
    MainLoop(tick)
}
